csrc/utils.h

```diff
@@ -843,4 +844,134 @@ using views::zip;
 
 #endif // C++23
 
+// Helper: turn T into reference_wrapper<U> if T is reference
+template <typename T>
+using Yielded = std::conditional_t<
+    std::is_reference_v<T>,
+    std::reference_wrapper<std::remove_reference_t<T>>,
+    T>;
+
+template <typename T>
+class Generator : public std::ranges::view_interface<Generator<T>> {
+ public:
+  struct promise_type;
+  using handle_type = std::coroutine_handle<promise_type>;
+  using stored_type = Yielded<T>;
+
+  Generator(handle_type h) : coro_(h) {}
+  Generator(Generator&& other) noexcept : coro_(other.coro_) {
+    other.coro_ = nullptr;
+  }
+  Generator& operator=(Generator&& other) noexcept {
+    if (this != &other) {
+      if (coro_) {
+        coro_.destroy();
+      }
+      coro_ = other.coro_;
+      other.coro_ = nullptr;
+    }
+    return *this;
+  }
+  ~Generator() {
+    if (coro_) {
+      coro_.destroy();
+    }
+  }
+  Generator(const Generator&) = delete;
+  Generator& operator=(const Generator&) = delete;
+
+  struct iterator {
+    using value_type = std::remove_reference_t<T>;
+    using reference = T;
+    using difference_type = std::ptrdiff_t;
+    using iterator_category = std::input_iterator_tag;
+
+    iterator() = default;
+    explicit iterator(handle_type h) : coro(h) {
+      ++(*this);
+    }
+
+    reference operator*() const {
+      if constexpr (std::is_reference_v<T>) {
+        return value->get(); // unwrap reference_wrapper<T>
+      } else {
+        return *value;
+      }
+    }
+
+    iterator& operator++() {
+      coro.resume();
+      if (coro.done()) {
+        if (coro.promise().exception) {
+          std::rethrow_exception(coro.promise().exception);
+        }
+        value.reset();
+      } else {
+        value = std::ref(coro.promise().current_value);
+      }
+      return *this;
+    }
+
+    iterator operator++(int) {
+      auto tmp = *this;
+      ++(*this);
+      return tmp;
+    }
+    bool operator==(std::default_sentinel_t) const {
+      return !value.has_value();
+    }
+    bool operator!=(std::default_sentinel_t) const {
+      return value.has_value();
+    }
+    friend bool operator==(std::default_sentinel_t s, const iterator& it) {
+      return it == s;
+    }
+    friend bool operator!=(std::default_sentinel_t s, const iterator& it) {
+      return it != s;
+    }
+
+    handle_type coro = nullptr;
+    std::optional<stored_type> value;
+  };
+
+  iterator begin() const {
+    return iterator{coro_};
+  }
+  std::default_sentinel_t end() const {
+    return {};
+  }
+
+ private:
+  handle_type coro_;
```

Comment:
https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

> Do not worry about saving horizontal space as it is far more important to make your code immediately understandable by a new reader. Minimize the use of abbreviations that would likely be unknown to someone outside your project (especially acronyms and initialisms).

Do you mean `coroutine_`?

Reply:
Yes, changed